src/CfgMgr32/CfgMgr32+CM_NOTIFY_EVENT_DATA.cs

```diff
@@ -0,0 +1,133 @@
+
+            /// <summary>
+            /// The GUID of the device interface class for the device interface to which the notification event data pertains.
+            /// </summary>
+            [FieldOffset(8)]
+            public Guid ClassGuid;
+
+            /// <summary>
+            /// The first character of the symbolic link path of the device interface to which the notification event data pertains.
+            /// Use <see cref="GetSymbolicLink(CM_NOTIFY_EVENT_DATA*, int)"/> to retrieve this value.
+            /// </summary>
+            [FieldOffset(24)]
+            public char SymbolicLink;
+
+            // DeviceHandle
+
+            /// <summary>
+            /// The GUID for the custom event.
+            /// </summary>
+            [FieldOffset(8)]
+            public Guid EventGuid;
+
+            /// <summary>
+            /// The offset of an optional string buffer. Usage depends on the contract for the EventGuid.
+            /// </summary>
+            [FieldOffset(24)]
+            public int NameOffset;
+
+            /// <summary>
+            /// The number of bytes that can be read from the Data member.
+            /// </summary>
+            [FieldOffset(28)]
+            public int DataSize;
+
+            /// <summary>
+            /// The first bit of optional binary data. Usage depends on the contract for the EventGuid.
+            /// </summary>
+            [FieldOffset(32)]
+            public byte Data;
+
+            // DeviceInstance
+
+            /// <summary>
+            /// The first character of the device instance ID of the device to which the notification event data pertains.
+            /// Use <see cref="GetInstanceId(CM_NOTIFY_EVENT_DATA*, int)"/> to retrieve this value.
+            /// </summary>
+            [FieldOffset(8)]
+            public char InstanceId;
+
+            /// <summary>
+            /// Gets the symbolic link path of the device interface to which the notification event data pertains.
+            /// </summary>
+            /// <param name="eventData">
+            /// The event notification.
+            /// </param>
+            /// <param name="eventDataSize">
+            /// The event notification size.
+            /// </param>
+            /// <returns>
+            /// The symbolic link path of the device interface to which the notification event data pertains
+            /// </returns>
+            public static string GetSymbolicLink(CM_NOTIFY_EVENT_DATA* eventData, int eventDataSize)
+            {
+                if (eventData->FilterType != CM_NOTIFY_FILTER_TYPE.CM_NOTIFY_FILTER_TYPE_DEVICEINTERFACE)
+                {
+                    throw new InvalidOperationException();
+                }
+
+                // Offset and count are represented in characters. Each character is 2 bytes wide.
+                // Trim the terminating \0 character at the end.
+                return new string((char*)eventData, 24 / 2, ((eventDataSize - 24) / 2) - 1);
```

Comment:
Avoid magic number `2` by using `sizeof(char)`.
I think you can avoid magic number `24` by using `&SymbolicLink` to point at the first character.

Reply:
&SymbolicLink doesn't seem work (it's a `static` method, but even making the method non-static doesn't help, as `this` gives you the struct, but not the address of the struct. I may have missed something, though.

I've worked around it by storing the offsets in `const`s, and referencing these consts in the `FieldOffset` attribute as well.

 Will make the `sizeof(char)` change.

Reply:
I pushed a commit to show how it's done. I also was able to avoid even the constants with the 'magic numbers'.